### PR TITLE
[Proposal] Separate execution config setup from tape preprocessing

### PR DESCRIPTION
**Pull request by albi3ro**
Currently, `Device.preprocess` has two roles:

1) transform and validate the provided tapes
2) Fill in the execution config with device specific choices.

This PR proposes to split these two responsibilities up.

Instead, we would have preprocess:
```
program = dev.preprocess(config)
new_tapes, postprocessing = program(tapes)

# Or
dev.preprocess(config)(tapes)
```

and config setup:
```
config = dev.setup_execution_config(tapes, original_config)
```

## Benefits

<b>Easier to write plugins</b>

Most devices will not to provide custom code for processing the config. They usually won't support derivatives, in which case they won't need to do anything to the device, or they will support "device" derivatives, in which default case behaviour will do everything they need. Only "power devices" like default qubit or lightning will need to be concerned with `setup_execution_config`.

<b>Better integrates `preprocess` with the transform infrastructure:</b>

Right now `preprocess` belongs in both to the transform infrastructure in addition to having this other config responsibility. It is not a "pure" transform ecosystem method.  If we make it a "pure" transform ecosystem method, we will gain additional functionality from the transform ecosystem.

For example, we could update the `TransformProgram` dispatcher onto different types of objects and unlock code like:
```
@hardware_dev.preprocess()
@qml.qnode(simulator_dev)
def circuit():
    return qml.probs()
```

The above workflow would be difficult to impossible with preprocessing also returning an execution config.

<b>Improved documentation:</b>

Currently, `preprocess` has to explain transforms, transform programs, transformations, and validation *and* config setup. This has left the config setup documentation lacking. By breaking it into a new method, we can have a documentation section focused purely on how to setup and interact with the execution config.

## Penalties

We are adding a new method.

We are once again having to change the Device API.  If we are ever going to do this, we need to do it now. But it might already be too late.

### CODE REVIEW

File: pennylane/devices/device_api.py [hunk 1]
```diff
@@ -199,10 +201,72 @@ def wires(self) -> Wires:
...
+
+        >>> new_config.device_options
+        {'max_workers': None, 'rng': Generator(PCG64) at 0x17D9C67A0, 'prng_key': None}
+
+        """
+        updated_values = {}
+        if (
+            execution_config.gradient_method in {"best", "device"}
+            and type(self).compute_derivatives != Device.compute_derivatives
+        ):
```
Comment (timmysilv) [inline pennylane/devices/device_api.py:256]: do we want this function to raise errors? what if the user asks for "device" but doesn't override `compute_derivates`?

File: tests/devices/qubit/test_preprocess.py [hunk 2]
```diff
@@ -594,24 +594,6 @@ def test_valid_tape_with_expansion(self, shots):
 class TestPreprocess:
     """Unit tests for ``qml.devices.qubit.preprocess``."""
 
-    def test_choose_best_gradient_method(self):
```
Comment (timmysilv) [inline tests/devices/qubit/test_preprocess.py:597]: should these tests be re-added elsewhere?

File: pennylane/devices/default_qubit.py [hunk 3]
```diff
@@ -205,18 +207,47 @@ def supports_derivatives(
...
+        updated_values["device_options"] = dict(execution_config.device_options)  # copy
+        if "max_workers" not in updated_values["device_options"]:
+            updated_values["device_options"]["max_workers"] = self._max_workers
+        if "rng" not in updated_values["device_options"]:
+            updated_values["device_options"]["rng"] = self._rng
+        if "prng_key" not in updated_values["device_options"]:
+            updated_values["device_options"]["prng_key"] = self._prng_key
+
+        print(updated_values)
+
```
Comment (timmysilv) [inline pennylane/devices/default_qubit.py:236]:
```suggestion
```

**Review by timmysilv — COMMENTED** [review]
I like the idea! Unsure about the name, but this makes sense. `preprocess` returning a config was definitely clumsy at times.

I do want to mention - we spoke in the past about devices providing a default config of sorts. Did you have a reason to shift in this direction, and could the ideas be worked into a single proposal?

File: tests/devices/qubit/test_preprocess.py (hunk 2 above)
Comment (albi3ro) [reply tests/devices/qubit/test_preprocess.py:597]: moved to just being in the device tests.

**Comment by albi3ro** [comment]
> I like the idea! Unsure about the name, but this makes sense. `preprocess` returning a config was definitely clumsy at times.
> 
> I do want to mention - we spoke in the past about devices providing a default config of sorts. Did you have a reason to shift in this direction, and could the ideas be worked into a single proposal?

Definitely fine with changing the name.

We could have it where `dev.setup_execution_config()` just returns the default config, where the tapes and initial config are empty.

**Comment by trbromley** [comment]
Shall we close this PR and reopen if we come back to this problem?

**Pull request closed without merging**